packages/back-end/src/routers/vercel-native-integration/vercel-native-integration.controller.ts

```diff
@@ -0,0 +1,531 @@
+import { createUser, getUserByEmail } from "back-end/src/models/UserModel";
+import { ReqContextClass } from "back-end/src/services/context";
+import { setResponseCookies } from "back-end/src/controllers/auth";
+import { OrganizationInterface } from "back-end/types/organization";
+import {
+  getVercelSSOToken,
+  syncVercelSdkWebhook,
+  deleteVercelSdkWebhook,
+} from "back-end/src/services/vercel-native-integration.service";
+import { createSDKConnection } from "back-end/src/models/SdkConnectionModel";
+import {
+  userAuthenticationValidator,
+  systemAuthenticationValidator,
+  UpsertInstallationPayload,
+  ProvisitionResource,
+  UpdateResource,
+  BillingPlan,
+} from "./vercel-native-integration.validators";
+
+const FREE_BILLING_PLAN: BillingPlan = {
+  description: "Free Billing Plan",
+  id: "free-billing-plan",
+  name: "Free",
+  type: "subscription",
+};
+
+const NON_FREE_BILLING_PLAN: BillingPlan = {
+  description: "Non Free Billing Plan",
+  id: "non-free",
+  name: "Non free",
+  type: "subscription",
+};
+
+const billingPlans = [FREE_BILLING_PLAN, NON_FREE_BILLING_PLAN] as const;
+
+const VERCEL_JKWS_URL = "https://marketplace.vercel.com/.well-known/jwks";
+
+const vercelJKWSKey = createRemoteJWKSet(new URL(VERCEL_JKWS_URL));
+
+const getBearerToken = (req: Request) => {
+  const authHeader = req.headers.authorization;
+
+  if (authHeader) {
+    const parts = authHeader.split(" ");
+    if (parts.length === 2 && parts[0].toLowerCase() === "bearer") {
+      return parts[1];
+    }
+  }
+
+  return undefined;
+};
+
+type CheckedAuth<T extends string | "user"> = T extends "user"
+  ? {
+      status: "authenticated";
+      authentication: z.infer<typeof userAuthenticationValidator>;
+    }
+  : {
+      status: "authenticated";
+      authentication: z.infer<typeof systemAuthenticationValidator>;
+    };
+
+const checkAuth = async <T extends string | "user">({
+  token,
+  type,
+}: {
+  token: string;
+  type: T;
+}): Promise<CheckedAuth<T> | { status: "error"; message: string }> => {
+  try {
+    const payload = await jwtVerify(token, await vercelJKWSKey);
+
+    if (type === "user")
+      return {
+        status: "authenticated",
+        authentication: userAuthenticationValidator.parse(payload),
+      } as CheckedAuth<T>;
+
+    if (type === "system")
+      return {
+        status: "authenticated",
+        authentication: systemAuthenticationValidator.parse(payload),
+      } as CheckedAuth<T>;
+
+    return { status: "error", message: `Unsupported authentication: ${type}` };
+  } catch (_err) {
+    return { status: "error", message: "Invalid credentials" };
+  }
+};
+
+const addOrganizationUser = (id: string, organization: OrganizationInterface) =>
+  updateOrganization(organization.id, {
+    members: [
+      ...organization.members,
+      {
+        id,
+        role: "admin",
+        dateCreated: new Date(),
+        limitAccessByEnvironment: false,
+        environments: [],
+      },
+    ],
+  });
```

Comment:
This is missing a bunch of stuff.  Better to use the `addMemberToOrg` method.  This is how we do it for other SSO connections:

```ts
  await addMemberToOrg({
    organization,
    userId,
    ...getDefaultRole(organization),
  });
  try {
    await sendNewMemberEmail(
      userName || "",
      userEmail || "",
      organization.name,
      organization.ownerEmail
    );
  } catch (e) {
    req.log.error(e, "Failed to send new member email");
  }
```